File: May-2023/Java/DesignUndergroundSystem.java
// 1396. Design Underground System
// Algorithm 
// find the difference between the time if both station
// add the time to the totalTime and divide by total count
class UndergroundSystem {
	Map<Integer, Pair<String, Integer>> checkIn;
	Map<String, Pair<Integer, Integer>> checkOut;

	public UndergroundSystem() {
		checkIn = new HashMap<>();
		checkOut = new HashMap<>();
	}

	public void checkIn(int id, String stationName, int t) {

		checkIn.put(id, new Pair<>(stationName, t));
	}

	public void checkOut(int id, String stationName, int t) {

		String entryStation = checkIn.get(id).getKey();
		int entryTime = checkIn.get(id).getValue();
		int totalTime = t - entryTime;
		String route = entryStation + "-" + stationName;

		checkOut.putIfAbsent(route, new Pair<>(0, 0));
		int totalT = checkOut.get(route).getKey() + totalTime;

		int count = checkOut.get(route).getValue() + 1;

		checkOut.put(route, new Pair<>(totalT, count));

		checkIn.remove(id);

	}

	public double getAverageTime(String startStation, String endStation) {
		return (double)(checkOut.get(startStation + "-" + endStation).getKey()) / (checkOut.get(startStation + "-" + endStation).getValue());
	}
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: May-2023/C++/DesignUnderGroundSystem.cxx
// 1396. Design Underground System
// Algorithm 
// find the difference between the time if both station
// add the time to the totalTime and divide by total count

class UndergroundSystem {
public:
    unordered_map<int, pair<string, int>> checkIn_mp;
    unordered_map<string, pair<int, int>> checkOut_mp;    
    
    UndergroundSystem() {
        
    }
    
    void checkIn(int id, string stationName, int t) {
        checkIn_mp[id] = {stationName, t};
    }
    
    void checkOut(int id, string stationName, int t) {
        auto& it                  = checkIn_mp[id];
        int entryTime             = it.second;
        int timeTaken             = t-entryTime;
        string entryStation       = it.first;
        string route              = entryStation + "_" + stationName;
        checkOut_mp[route].first  += timeTaken;
        checkOut_mp[route].second += 1;
        checkIn_mp.erase(id); 
    }
    
    double getAverageTime(string startStation, string endStation) {
        string route =  startStation + "_" + endStation;
        auto&it      = checkOut_mp[route];
        return (double)it.first/it.second;
    }
};
